File: frontends/exchange-os/src/lib/troptions/genius/gates.ts
import type {
  ComplianceGate,
  Decision,
  GateRequirementLevel,
  IssuerMode,
  NamespaceRecord,
  StablecoinAction,
  TroptionsGeniusProfile,
} from "@/lib/troptions/genius/types";

const LIVE_ACTIONS: StablecoinAction[] = ["live_mint", "live_burn", "live_redemption"];

const LIVE_PROFILE_REQUIREMENTS: Array<keyof TroptionsGeniusProfile> = [
  "legalCounselMemoApproved",
  "regulatorApprovalRecorded",
  "boardApprovalRecorded",
  "reservePolicyApproved",
  "reserveAttestationCurrent",
  "redemptionPolicyApproved",
  "amlSanctionsProgramApproved",
  "kycKybProviderActive",
  "incidentResponsePlanApproved",
  "consumerDisclosuresApproved",
  "chainRiskReviewApproved",
  "smartContractAuditApproved",
];

const LIVE_GATE_IDS = [
  "legal-counsel-memo",
  "board-approval",
  "regulator-approval",
  "reserve-policy",
  "reserve-attestation",
  "redemption-policy",
  "aml-bsa-policy",
  "sanctions-policy",
  "kyc-kyb-provider",
  "chain-analytics-provider",
  "smart-contract-audit",
  "incident-response-plan",
  "consumer-disclosures",
];

const PARTNER_OR_LICENSED: IssuerMode[] = ["partnered_ppsi", "licensed_ppsi"];

function gateSatisfiesRequirement(gate: ComplianceGate, level: GateRequirementLevel): boolean {
  const ordered: GateRequirementLevel[] = ["sandbox", "partner_ready", "regulator_ready", "live"];
  return ordered.indexOf(gate.requiredFor) <= ordered.indexOf(level);
}

export function listBlockingItems(gates: ComplianceGate[], requestedStatus: GateRequirementLevel): string[] {
  return gates
    .filter((gate) => gateSatisfiesRequirement(gate, requestedStatus))
    .filter((gate) => gate.status !== "approved")
    .map((gate) => `${gate.label} (${gate.status})`);
}

export function evaluateNamespaceSettlement(namespace: NamespaceRecord, profile: TroptionsGeniusProfile): Decision {
  const reasons: string[] = [];
  const requiredNextSteps: string[] = [];

  if (![namespace.kycStatus, namespace.kybStatus].includes("mock_approved")) {
    reasons.push("Namespace identity controls are not mock-approved for sandbox activity.");
    requiredNextSteps.push("Mock-approve KYC and KYB before settlement simulation.");
  }

  if (namespace.sanctionsStatus !== "clear") {
    reasons.push("Namespace sanctions status is not clear.");
    requiredNextSteps.push("Resolve sanctions review before any settlement path.");
  }

  return {
    allowed: reasons.length === 0,
    status: reasons.length === 0 ? "sandbox_ready" : "blocked",
    reasons,
    requiredNextSteps,
    riskRating: reasons.length === 0 ? "low" : "high",
  };
}

export function evaluateStablecoinAction(
  profile: TroptionsGeniusProfile,
  gates: ComplianceGate[],
  action: StablecoinAction,
  namespace?: NamespaceRecord,
): Decision {
  const reasons: string[] = [];
  const requiredNextSteps: string[] = [];

  if (LIVE_ACTIONS.includes(action)) {
    if (!PARTNER_OR_LICENSED.includes(profile.issuerMode)) {
      reasons.push("Live stablecoin actions require either a licensed PPSI or an approved partnered PPSI structure.");
    }

    const missingProfileFlags = LIVE_PROFILE_REQUIREMENTS.filter((field) => !profile[field]);
    if (missingProfileFlags.length > 0) {
      reasons.push(`Required live approvals are missing: ${missingProfileFlags.join(", ")}.`);
      requiredNextSteps.push("Complete all live approval flags before requesting live activation.");
    }

    const gateBlockers = listBlockingItems(gates.filter((gate) => LIVE_GATE_IDS.includes(gate.id)), "live");
    if (gateBlockers.length > 0) {
      reasons.push(`Live gate blockers remain: ${gateBlockers.join("; ")}.`);
      requiredNextSteps.push("Move all live-required gates to approved status.");
    }
  }

  if (!LIVE_ACTIONS.includes(action)) {
    const sandboxBlockers = listBlockingItems(gates, "sandbox");
    if (["simulate_mint", "simulate_burn", "simulate_transfer", "simulate_redemption_request"].includes(action) && sandboxBlockers.length > 0) {
      reasons.push(`Sandbox readiness is incomplete: ${sandboxBlockers.join("; ")}.`);
      requiredNextSteps.push("Approve all sandbox-required gates before running simulations.");
    }
  }

  if (namespace) {
    const namespaceDecision = evaluateNamespaceSettlement(namespace, profile);
    if (!namespaceDecision.allowed) {
      reasons.push(...namespaceDecision.reasons);
      requiredNextSteps.push(...namespaceDecision.requiredNextSteps);
    }
  }

  if (LIVE_ACTIONS.includes(action) && profile.publicChainAllowed && !profile.chainRiskReviewApproved) {
    reasons.push("Public chain use does not bypass chain risk review and smart contract audit requirements.");
  }

  if (reasons.length === 0) {
    return {
      allowed: true,
      status: LIVE_ACTIONS.includes(action) ? "live_enabled" : "sandbox_ready",
      reasons: [LIVE_ACTIONS.includes(action) ? "All live requirements are approved." : "Sandbox-only action approved."],
      requiredNextSteps: [],
      riskRating: LIVE_ACTIONS.includes(action) ? "medium" : "low",
    };
  }

  return {
    allowed: false,
    status: LIVE_ACTIONS.includes(action) ? "blocked_live" : "blocked_sandbox",
    reasons,
    requiredNextSteps: Array.from(new Set(requiredNextSteps)),
    riskRating: LIVE_ACTIONS.includes(action) ? "critical" : "high",
  };
}

File: frontends/exchange-os/troptions-rust-l1/crates/troptions-genius-core/src/gates.rs
use crate::models::{
    ComplianceGate, Decision, GateStatus, IssuerMode, RequirementLevel, StablecoinAction,
    TroptionsGeniusProfile,
};

fn requirement_order(level: RequirementLevel) -> u8 {
    match level {
        RequirementLevel::Sandbox => 0,
        RequirementLevel::PartnerReady => 1,
        RequirementLevel::RegulatorReady => 2,
        RequirementLevel::Live => 3,
    }
}

fn is_live_action(action: StablecoinAction) -> bool {
    matches!(
        action,
        StablecoinAction::LiveMint | StablecoinAction::LiveBurn | StablecoinAction::LiveRedemption
    )
}

pub fn list_blocking_items(
    gates: &[ComplianceGate],
    requested_status: RequirementLevel,
) -> Vec<String> {
    gates
        .iter()
        .filter(|gate| requirement_order(gate.required_for) <= requirement_order(requested_status))
        .filter(|gate| gate.status != GateStatus::Approved)
        .map(|gate| format!("{} ({:?})", gate.label, gate.status).to_lowercase())
        .collect()
}

pub fn evaluate_stablecoin_action(
    profile: &TroptionsGeniusProfile,
    gates: &[ComplianceGate],
    action: StablecoinAction,
) -> Decision {
    let mut reasons = Vec::new();
    let mut required_next_steps = Vec::new();

    if is_live_action(action) {
        if !matches!(
            profile.issuer_mode,
            IssuerMode::LicensedPpsi | IssuerMode::PartneredPpsi
        ) {
            reasons.push(
                "live stablecoin actions require a licensed or partnered ppsi path".to_string(),
            );
        }

        if !profile.legal_counsel_memo_approved {
            reasons.push("legal counsel memo not approved".to_string());
        }
        if !profile.regulator_approval_recorded {
            reasons.push("regulator approval not recorded".to_string());
        }
        if !profile.board_approval_recorded {
            reasons.push("board approval not recorded".to_string());
        }
        if !profile.reserve_policy_approved {
            reasons.push("reserve policy not approved".to_string());
        }
        if !profile.reserve_attestation_current {
            reasons.push("reserve attestation is not current".to_string());
        }
        if !profile.redemption_policy_approved {
            reasons.push("redemption policy not approved".to_string());
        }
        if !profile.aml_sanctions_program_approved {
            reasons.push("aml and sanctions program not approved".to_string());
        }
        if !profile.kyc_kyb_provider_active {
            reasons.push("kyc/kyb provider is not active".to_string());
        }
        if !profile.incident_response_plan_approved {
            reasons.push("incident response plan not approved".to_string());
        }
        if !profile.consumer_disclosures_approved {
            reasons.push("consumer disclosures not approved".to_string());
        }
        if !profile.chain_risk_review_approved {
            reasons.push("chain risk review not approved".to_string());
        }
        if !profile.smart_contract_audit_approved {
            reasons.push("smart contract audit not approved".to_string());
        }

        let live_blockers = list_blocking_items(gates, RequirementLevel::Live);
        if !live_blockers.is_empty() {
            reasons.push(format!(
                "live blockers remain: {}",
                live_blockers.join("; ")
            ));
        }

        if profile.public_chain_allowed && !profile.chain_risk_review_approved {
            reasons.push("public chain allowance does not bypass chain risk review".to_string());
        }

        if !reasons.is_empty() {
            required_next_steps
                .push("complete live gate approvals before requesting activation".to_string());
        }
    } else {
        let sandbox_blockers = list_blocking_items(gates, RequirementLevel::Sandbox);
        if matches!(
            action,
            StablecoinAction::SimulateMint
                | StablecoinAction::SimulateBurn
                | StablecoinAction::SimulateTransfer
                | StablecoinAction::SimulateRedemptionRequest
        ) && !sandbox_blockers.is_empty()
        {
            reasons.push(format!(
                "sandbox readiness incomplete: {}",
                sandbox_blockers.join("; ")
            ));
            required_next_steps.push("approve sandbox gates before simulation".to_string());
        }
    }

    if reasons.is_empty() {
        Decision {
            allowed: true,
            status: if is_live_action(action) {
                "live_enabled".to_string()
            } else {
                "sandbox_ready".to_string()
            },
            reasons: vec![if is_live_action(action) {
                "all live requirements approved".to_string()
            } else {
                "sandbox-only action approved".to_string()
            }],
            required_next_steps,
        }
    } else {
        Decision {
            allowed: false,
            status: if is_live_action(action) {
                "blocked_live".to_string()
            } else {
                "blocked_sandbox".to_string()
            },
            reasons,
            required_next_steps,
        }
    }
}
